### PR TITLE
8277137: Set OnSpinWaitInst/OnSpinWaitInstCount defaults to "isb"/1 for Arm Neoverse N1

### PR DESCRIPTION
**Pull request by eastig**
One `ISB` implementation of `Thread.OnSpinWait` shows performance improvements on Graviton2 (Arm Neoverse N1 implementation), e.g. https://github.com/openjdk/jdk/pull/5562#issuecomment-966153163. 

Testing:
- `make test TEST=gtest`: Passed
- `make run-test TEST=tier1`: Passed
- `make run-test TEST=tier2`: Passed
- `make run-test TEST=hotspot/jtreg/compiler/onSpinWait`: Passed

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277137](https://bugs.openjdk.java.net/browse/JDK-8277137): Set OnSpinWaitInst/OnSpinWaitInstCount defaults to "isb"/1 for Arm Neoverse N1

### Reviewers
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6415/head:pull/6415` \
`$ git checkout pull/6415`

Update a local copy of the PR: \
`$ git checkout pull/6415` \
`$ git pull https://git.openjdk.java.net/jdk pull/6415/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6415`

View PR using the GUI difftool: \
`$ git pr show -t 6415`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6415.diff">https://git.openjdk.java.net/jdk/pull/6415.diff</a>

</details>

### CODE REVIEW

**Review by phohensee — CHANGES_REQUESTED** [review]
I'd explicitly set OnSpinWaitInstCount because one has to go find the default value in another file to understand what's going to happen. So I'd add:

FLAG_SET_DEFAULT(OnSpinWaitInstCount, 1);

**Comment by eastig** [comment]
> I'd explicitly set OnSpinWaitInstCount because one has to go find the default value in another file to understand what's going to happen. So I'd add:
> 
> FLAG_SET_DEFAULT(OnSpinWaitInstCount, 1);

Thank you for reviewing.
Done.

File: src/hotspot/cpu/aarch64/vm_version_aarch64.cpp [hunk 1]
```diff
@@ -202,6 +202,11 @@ void VM_Version::initialize() {
     if (FLAG_IS_DEFAULT(UseSIMDForMemoryOps)) {
       FLAG_SET_DEFAULT(UseSIMDForMemoryOps, true);
     }
+
+    if (FLAG_IS_DEFAULT(OnSpinWaitInst) && FLAG_IS_DEFAULT(OnSpinWaitInstCount)) {
```
Comment (nick-arm) [inline src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:206]: Should these two be set independently? If I pass `-XX:OnSpinWaitInstCount=2` then `OnSpinWaitInst` will default to "none".

Comment (eastig) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:206]: Hi Nick,
Thank you for reviewing the PR.

> Should these two be set independently?

I don't mind.

Comment (eastig) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.cpp]: Done.

**Comment by theRealAph** [comment]
Did we establish that this is the right default for Neoverse N1? I know that we've found a benchmark where it's a win, bit I'm not sure that's the same thing. On the other hand, do we know of possible cases where ISB makes things worse?

**Comment by eastig** [comment]
Hi Andrew,
Thank you for reviewing.

> Did we establish that this is the right default for Neoverse N1?

This is based on:
- MySql: https://bugs.mysql.com/bug.php?id=100664
- MongoDB: https://jira.mongodb.org/browse/WT-6872
- Netty: https://github.com/netty/netty/pull/11677
- Customers' benchmarks and workloads.
- Experiments with two and three `ISB` instructions.

> On the other hand, do we know of possible cases where ISB makes things worse?

`Thread.onSpinWait` makes things worse when synchronisation overhead is not on the critical path. It might not improve performance when there is thread contention. In this case it might not give CPU resources to another thread. This applies to both arm64 and x86_64.
For example, my x86 system:
```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              16
On-line CPU(s) list: 0-15
Thread(s) per core:  2
Core(s) per socket:  8
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
Stepping:            7
CPU MHz:             3097.588
BogoMIPS:            4999.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
```
Results of `org.openjdk.bench.java.lang.ThreadOnSpinWaitSharedCounter` with 4 threads running on 2 vCPUs:
- `taskset -c 0-1 build/linux-x86_64-server-release/images/jdk/bin/java -XX:+UnlockDiagnosticVMOptions -XX:DisableIntrinsic=_onSpinWait -jar build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 3 org.openjdk.bench.java.lang.ThreadOnSpinWaitSharedCounter` 
```
Benchmark                            (maxNum)  (threadCount)  Mode  Cnt   Score   Error  Units
ThreadOnSpinWaitSharedCounter.trial   1000000              4  avgt   15  45.317 ± 1.741  ms/op
```
- `taskset -c 0-1 build/linux-x86_64-server-release/images/jdk/bin/java -jar build/linux-x86_64-server-release/images/test/micro/benchmarks.jar -f 3 org.openjdk.bench.java.lang.ThreadOnSpinWaitSharedCounter`
```
Benchmark                            (maxNum)  (threadCount)  Mode  Cnt   Score   Error  Units
ThreadOnSpinWaitSharedCounter.trial   1000000              4  avgt   15  55.530 ± 4.606  ms/op
```

X86 `PAUSE` based implementation causes 22.5% slowdown.

**Review by theRealAph — APPROVED** [review]

**Comment by eastig** [comment]
@theRealAph Thank you.

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by eastig** [comment]
@phohensee Thank you.

**Review by nick-arm — APPROVED** [review]

**Pull request closed without merging**